receiver/receivercreator/receiver.go

```diff
@@ -0,0 +1,149 @@
+	if nextConsumer == nil {
+		return nil, errNilNextConsumer
+	}
+
+	r := &receiverCreator{
+		logger:       logger,
+		nextConsumer: nextConsumer,
+		cfg:          cfg,
+	}
+	return r, nil
+}
+
+// loadRuntimeReceiverConfig loads the given subreceiverConfig merged with config values
+// that may have been discovered at runtime.
+func (rc *receiverCreator) loadRuntimeReceiverConfig(factory component.ReceiverFactoryOld,
+	staticSubConfig *subreceiverConfig, discoveredConfig userConfigMap) (configmodels.Receiver, error) {
+	mergedConfig := config.NewViper()
+
+	// Merge in the config values specified in the config file.
+	if err := mergedConfig.MergeConfigMap(staticSubConfig.config); err != nil {
+		return nil, fmt.Errorf("failed to merge subreceiver config from config file: %v", err)
+	}
+
+	// Merge in discoveredConfig containing values discovered at runtime.
+	if err := mergedConfig.MergeConfigMap(discoveredConfig); err != nil {
+		return nil, fmt.Errorf("failed to merge subreceiver config from discovered runtime values: %v", err)
+	}
+
+	viperConfig := config.NewViper()
+
+	// Load config under <receiver>/<id> since loadReceiver and CustomUnmarshaler expects this structure.
+	viperConfig.Set(staticSubConfig.fullName, mergedConfig.AllSettings())
+
+	receiverConfig, err := config.LoadReceiver(staticSubConfig.fullName, viperConfig, map[string]component.ReceiverFactoryBase{
+		staticSubConfig.receiverType: factory,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to load subreceiver config: %v", err)
+	}
+	// Sets dynamically created receiver to something like receiver_creator/1/redis{endpoint="localhost:6380"}.
+	// TODO: Need to make sure this is unique (just endpoint is probably not totally sufficient).
+	receiverConfig.SetName(fmt.Sprintf("%s/%s{endpoint=%q}", rc.cfg.Name(), staticSubConfig.fullName, mergedConfig.GetString("endpoint")))
+	return receiverConfig, nil
+}
+
+// createRuntimeReceiver creates a receiver that is discovered at runtime.
+func (rc *receiverCreator) createRuntimeReceiver(factory component.ReceiverFactoryOld, cfg configmodels.Receiver) (component.MetricsReceiver, error) {
+	return factory.CreateMetricsReceiver(rc.logger, cfg, rc.nextConsumer)
+}
+
+// Start receiver_creator.
+func (rc *receiverCreator) Start(host component.Host) error {
+	// TODO: Temporarily load a single instance of all subreceivers for testing.
+	// Will be done in reaction to observer events later.
+	for _, subconfig := range rc.cfg.subreceiverConfigs {
+		factory := host.GetFactory(component.KindReceiver, subconfig.receiverType)
+
+		if factory == nil {
+			rc.logger.Error("unable to lookup factory for receiver", zap.String("receiver", subconfig.receiverType))
+			continue
+		}
+
+		receiverFactory := factory.(component.ReceiverFactoryOld)
+
+		cfg, err := rc.loadRuntimeReceiverConfig(receiverFactory, subconfig, userConfigMap{})
```

Comment:
I am confused: where is this `userConfigMap{}` coming from? Can we just remove from the moment?

Reply:
It's how the endpoint will be passed when an instance is discovered. It will be used in the next PR that I have in progress.